base/dates/io.jl

```diff
@@ -152,7 +152,7 @@ function parse(x::AbstractString,df::DateFormat)
         throw(ArgumentError("Delimiter mismatch. Couldn't find first delimiter, \"$(df.slots[1].transition)\", in date string"))
     end
     periods = Period[]
-    extra = []
+    extra = Period[]
```

Comment:
@omus, that's an improvement, for sure. Although it doesn't solve the problem for `@code_warntype`. Also, `eltype` returns `Any` for the original test case:
```julia
julia> eltype(Date("2015-01-01",Dates.DateFormat("yyyy-mm-dd")))
Any
```
See full log at https://gist.github.com/felipenoris/10334d4f09549f928e6e.
Maybe this is as far as we can go without doing a breaking change.